### PR TITLE
+str Add `startAfterNrOfConsumers` to BroadcastHub.

### LINKED ISSUE #28327: BroadcastHub doesn't wait for first materialization of Source
This is what _Akka Streams_ documentation states about `Source` materialized from `BroadcastHub` `Sink` ([link][1]):

> The resulting Source can be materialized any number of times, each materialization effectively attaching a new subscriber. If there are no subscribers attached to this hub then it will not drop any elements but instead backpressure the upstream producer until subscribers arrive.

Therefore I would expect that this:

```java
 Source<String, NotUsed> bcast =
            Source.from(Arrays.asList("1", "2", "3"))
                  .log("before broadcasting")
                  .runWith(BroadcastHub.of(String.class), materializer);

 bcast.log("after broadcasting")
      .runWith(Sink.ignore(), materializer);
```

would effectively backpressure the upstream of the `BroadcastHub` `Sink` until the `bcast` `Source` is materialized at least once, but it is not the case. If I run the above, the output is:

```
[before broadcasting] Element: 1
[before broadcasting] Element: 2
[before broadcasting] Element: 3
[before broadcasting] Upstream finished.
[after broadcasting] Upstream finished.
```

which is contrary to the documentation, as the elements are effectively dropped and never broadcasted. If I limit the buffer of the `Broadcast` `Sink` like this:

```
        Source<String, NotUsed> bcast =
            Source.from(Arrays.asList("1", "2", "3"))
                  .log("before broadcasting")
                  .runWith(BroadcastHub.of(String.class, 2), materializer); //buffer smaller than number of events

```

then all the elements are broadcasted:

```
[before broadcasting] Element: 1
[before broadcasting] Element: 2
[after broadcasting] Element: 1
[after broadcasting] Element: 2
[before broadcasting] Element: 3
[before broadcasting] Upstream finished.
[after broadcasting] Element: 3
[after broadcasting] Upstream finished.
```

The way I see it it's either a bug or the documentation is misleading?

originally posted at SO: 

https://stackoverflow.com/questions/58974630/broadcasthub-doesnt-wait-for-first-materialization-of-source

  [1]: https://doc.akka.io/docs/akka/current/stream/stream-dynamic.html

### PR DESCRIPTION
**Pull request by He-Pin**
refs: https://github.com/akka/akka/issues/27163

And can fix https://github.com/akka/akka/issues/28327 if we set the default `startAfterNrOfConsumers` to `1`, but may come with a separate PR.

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala [hunk 1]
```diff
@@ -515,8 +545,29 @@ private[akka] class BroadcastHub[T](bufferSize: Int)
...
+          addConsumer(consumer, newOffset)
+          checkUnblock(previousOffset)
+        case NeedWakeup(id, previousOffset, currentOffset) =>
+          // Move the consumer from its last known offset to its new one. Check if we are unblocked.
+          val consumer = findAndRemoveConsumer(id, previousOffset)
+          addConsumer(consumer, currentOffset)
+
+          // Also check if the consumer is now unblocked since we published an element since it went asleep.
+          if (currentOffset != tail) consumer.callback.invoke(Wakeup)
+          checkUnblock(previousOffset)
```
Comment (He-Pin) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:569]: I just move this two case up for better performance.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala [hunk 2]
```diff
@@ -547,24 +602,9 @@ private[akka] class BroadcastHub[T](bufferSize: Int)
                 n += 1
               }
               head = finalOffset
-              if (!hasBeenPulled(in)) pull(in)
+              tryPull()
```
Comment (He-Pin) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:605]: I think this is needed, if we set the `startAfterNrOfConsumers` to `2` and register `1`  consumer and then unregister it, but not have a test to cover this, can add one.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala [hunk 3]
```diff
@@ -1226,20 +1269,20 @@ object PartitionHub {
...
-            }
-
             consumer.callback.invoke(Initialize)
+          }
 
-            if (initialized && pending.nonEmpty) {
-              pending.foreach(publish)
-              pending = Vector.empty[T]
-            }
+          if (consumerInfo.size >= startAfterNrOfConsumers) {
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:1275]: yes, looks correct to move the following outside of the foreach

File: akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala [hunk 4]
```diff
@@ -416,6 +416,31 @@ object BroadcastHub {
...
+   * cancelled are simply removed from the dynamic set of consumers.
+   *
+   * @param startAfterNrOfConsumers Elements are buffered until this number of consumers have been connected.
+   *   This is only used initially when the operator is starting up, i.e. it is not honored when consumers have
+   *   been removed (canceled).
+   * @param bufferSize Buffer size used by the producer. Gives an upper bound on how "far" from each other two
+   *                   concurrent consumers can be in terms of element. If this buffer is full, the producer
+   *                   is backpressured. Must be a power of two and less than 4096.
+   */
+  def sink[T](startAfterNrOfConsumers: Int, bufferSize: Int): Sink[T, Source[T, NotUsed]] =
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:441]: same in javadsl?

**Review by patriknw — COMMENTED** [review]
looking good, very similar to what we have in partition hub

**Comment by He-Pin** [comment]
> looking good, very similar to what we have in partition hub

Yes, I took a look at what you did in the PartionHub.

**Review by patriknw — APPROVED** [review]
LGTM, thanks

**Review by johanandren — APPROVED** [review]

**Pull request merged**